### PR TITLE
Fix staticcheck failures for vendor/k8s.io/client-go/rest/*

### PR DESCRIPTION
**Pull request by JaredTan95**
Signed-off-by: JaredTan95 <jian.tan@daocloud.io>

#### What this PR does / why we need it:
Part of  https://github.com/kubernetes/kubernetes/issues/92402

```bash
vendor/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:482:9: assigning the result of this type assertion to a variable (switch err := err.(type)) could eliminate type assertions in switch cases (S1034)
        vendor/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:497:8: could eliminate this type assertion
vendor/k8s.io/client-go/plugin/pkg/client/auth/openstack/openstack_stub.go:33:14: error strings should not be capitalized (ST1005)
vendor/k8s.io/client-go/rest/client_test.go:376:96: the argument is already a string, there's no need to use fmt.Sprintf (S1025)
vendor/k8s.io/client-go/rest/config.go:260:8: should convert c (type TLSClientConfig) to sanitizedTLSClientConfig instead of using struct literal (S1016)
vendor/k8s.io/client-go/rest/config_test.go:283:5: error var fakeDialerError should have name of the form errFoo (ST1012)
vendor/k8s.io/client-go/rest/config_test.go:295:5: error var fakeAuthProviderConfigPersisterError should have name of the form errFoo (ST1012)
vendor/k8s.io/client-go/rest/connection_test.go:65:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
        vendor/k8s.io/client-go/rest/connection_test.go:55:3: call to T.Fatalf
vendor/k8s.io/client-go/rest/connection_test.go:112:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
        vendor/k8s.io/client-go/rest/connection_test.go:62:3: call to T.Fatalf
vendor/k8s.io/client-go/rest/connection_test.go:145:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
        vendor/k8s.io/client-go/rest/connection_test.go:62:3: call to T.Fatalf
vendor/k8s.io/client-go/rest/exec.go:24:2: package "k8s.io/client-go/pkg/apis/clientauthentication" is being imported more than once (ST1019)
        vendor/k8s.io/client-go/rest/exec.go:25:2: other import of "k8s.io/client-go/pkg/apis/clientauthentication"
vendor/k8s.io/client-go/rest/exec_test.go:31:2: package "k8s.io/client-go/pkg/apis/clientauthentication" is being imported more than once (ST1019)
        vendor/k8s.io/client-go/rest/exec_test.go:32:2: other import of "k8s.io/client-go/pkg/apis/clientauthentication"
vendor/k8s.io/client-go/rest/request.go:515:32: possible nil pointer dereference (SA5011)
        vendor/k8s.io/client-go/rest/request.go:519:5: this check suggests that the pointer can be nil
vendor/k8s.io/client-go/rest/request.go:515:32: possible nil pointer dereference (SA5011)
vendor/k8s.io/client-go/rest/request.go:519:5: this check suggests that the pointer can be nil
vendor/k8s.io/client-go/rest/request_test.go:927:6: type errorReader is unused (U1000)
vendor/k8s.io/client-go/rest/request_test.go:931:22: func errorReader.Read is unused (U1000)
vendor/k8s.io/client-go/rest/request_test.go:932:22: func errorReader.Close is unused (U1000)
vendor/k8s.io/client-go/rest/request_test.go:1190:5: should use 'return err.Error() == "a container name must be specified for pod kube-dns-v20-mz5cv, choose one of: [kubedns dnsmasq healthz]"' instead of 'if err.Error() == "a container name must be specified for pod kube-dns-v20-mz5cv, choose one of: [kubedns dnsmasq healthz]" { return true }; return false' (S1008)
vendor/k8s.io/client-go/rest/request_test.go:1216:6: type fakeUpgradeConnection is unused (U1000)
vendor/k8s.io/client-go/rest/request_test.go:1218:33: func (*fakeUpgradeConnection).CreateStream is unused (U1000)
vendor/k8s.io/client-go/rest/request_test.go:1221:33: func (*fakeUpgradeConnection).Close is unused (U1000)
vendor/k8s.io/client-go/rest/request_test.go:1224:33: func (*fakeUpgradeConnection).CloseChan is unused (U1000)
vendor/k8s.io/client-go/rest/request_test.go:1227:33: func (*fakeUpgradeConnection).SetIdleTimeout is unused (U1000)
vendor/k8s.io/client-go/rest/request_test.go:1230:6: type fakeUpgradeRoundTripper is unused (U1000)
vendor/k8s.io/client-go/rest/request_test.go:1235:35: func (*fakeUpgradeRoundTripper).RoundTrip is unused (U1000)
vendor/k8s.io/client-go/rest/request_test.go:1246:35: func (*fakeUpgradeRoundTripper).NewConnection is unused (U1000)
vendor/k8s.io/client-go/rest/request_test.go:1335:3: redundant return statement (S1023)
vendor/k8s.io/client-go/rest/urlbackoff_test.go:69:3: var backoffSec is of type time.Duration; don't use unit-specific suffix "Sec" (ST1011)
vendor/k8s.io/client-go/rest/watch/decoder_test.go:54:3: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
        vendor/k8s.io/client-go/rest/watch/decoder_test.go:57:5: call to T.Fatalf
vendor/k8s.io/client-go/rest/watch/decoder_test.go:70:3: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
        vendor/k8s.io/client-go/rest/watch/decoder_test.go:73:5: call to T.Fatalf
vendor/k8s.io/client-go/testing/fixture.go:513:2: should use 'return !(!resourceCovers)' instead of 'if !resourceCovers { return false }; return true' (S1008)
vendor/k8s.io/client-go/testing/fixture.go:534:2: should use 'return !(!resourceCovers)' instead of 'if !resourceCovers { return false }; return true' (S1008)
vendor/k8s.io/client-go/testing/fixture.go:555:2: should use 'return !(!resourceCovers)' instead of 'if !resourceCovers { return false }; return true' (S1008)
vendor/k8s.io/client-go/third_party/forked/golang/template/funcs.go:55:5: var builtinFuncs is unused (U1000)
vendor/k8s.io/client-go/third_party/forked/golang/template/funcs.go:80:6: func addFuncs is unused (U1000)
vendor/k8s.io/client-go/third_party/forked/golang/template/funcs.go:99:6: func findFunction is unused (U1000)
vendor/k8s.io/client-go/third_party/forked/golang/template/funcs.go:472:5: should use !strings.ContainsAny(s, `'"&<>`) instead (S1003)
vendor/k8s.io/client-go/tools/cache/delta_fifo.go:432:10: error strings should not be capitalized (ST1005)
vendor/k8s.io/client-go/tools/cache/fifo.go:266:2: should use 'return f.closed' instead of 'if f.closed { return true }; return false' (S1008)
vendor/k8s.io/client-go/tools/cache/heap.go:307:2: should use 'return h.closed' instead of 'if h.closed { return true }; return false' (S1008)
vendor/k8s.io/client-go/tools/cache/reflector.go:234:23: error strings should not be capitalized (ST1005)
vendor/k8s.io/client-go/tools/clientcmd/config.go:138:2: should use 'return len(o.LoadingRules.ExplicitPath) > 0' instead of 'if len(o.LoadingRules.ExplicitPath) > 0 { return true }; return false' (S1008)
vendor/k8s.io/client-go/tools/clientcmd/helpers.go:35:12: error strings should not be capitalized (ST1005)
vendor/k8s.io/client-go/tools/clientcmd/validation.go:237:47: error strings should not end with punctuation or a newline (ST1005)
vendor/k8s.io/client-go/tools/clientcmd/validation.go:267:48: error strings should not end with punctuation or a newline (ST1005)
vendor/k8s.io/client-go/tools/clientcmd/validation.go:275:48: error strings should not end with punctuation or a newline (ST1005)
vendor/k8s.io/client-go/tools/clientcmd/validation_test.go:577:7: receiver name should not be an underscore, omit the name if it is unused (ST1006)
vendor/k8s.io/client-go/tools/clientcmd/validation_test.go:581:7: receiver name should not be an underscore, omit the name if it is unused (ST1006)
vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:101:15: error strings should not be capitalized (ST1005)
vendor/k8s.io/client-go/tools/leaderelection/leaderelection.go:101:15: error strings should not end with punctuation or a newline (ST1005)
vendor/k8s.io/client-go/tools/leaderelection/metrics.go:75:7: receiver name should not be an underscore, omit the name if it is unused (ST1006)
vendor/k8s.io/client-go/tools/leaderelection/resourcelock/interface.go:143:15: error strings should not be capitalized (ST1005)
vendor/k8s.io/client-go/tools/remotecommand/remotecommand_test.go:93:2: should use a simple channel send/receive instead of select with a single case (S1000)
vendor/k8s.io/client-go/tools/remotecommand/v1.go:116:17: error strings should not be capitalized (ST1005)
vendor/k8s.io/client-go/tools/remotecommand/v1.go:120:17: error strings should not be capitalized (ST1005)
vendor/k8s.io/client-go/tools/watch/retrywatcher.go:81:47: should use make(chan watch.Event) instead (S1019)
vendor/k8s.io/client-go/tools/watch/retrywatcher_test.go:125:35: should use make(chan watch.Event) instead (S1019)
vendor/k8s.io/client-go/transport/round_trippers_test.go:449:23: unnecessary use of fmt.Sprintf (S1039)
vendor/k8s.io/client-go/transport/round_trippers_test.go:464:23: unnecessary use of fmt.Sprintf (S1039)
vendor/k8s.io/client-go/transport/round_trippers_test.go:483:34: unnecessary use of fmt.Sprintf (S1039)
vendor/k8s.io/client-go/util/cert/io.go:33:5: should omit comparison to bool constant, can be simplified to !certReadable (S1002)
vendor/k8s.io/client-go/util/cert/io.go:33:30: should omit comparison to bool constant, can be simplified to !keyReadable (S1002)
vendor/k8s.io/client-go/util/cert/io.go:37:5: should omit comparison to bool constant, can be simplified to !certReadable (S1002)
vendor/k8s.io/client-go/util/cert/io.go:41:5: should omit comparison to bool constant, can be simplified to !keyReadable (S1002)
vendor/k8s.io/client-go/util/certificate/certificate_manager.go:412:27: error strings should not be capitalized (ST1005)
vendor/k8s.io/client-go/util/certificate/csr/csr.go:32:2: package "k8s.io/apimachinery/pkg/api/errors" is being imported more than once (ST1019)
        vendor/k8s.io/client-go/util/certificate/csr/csr.go:33:2: other import of "k8s.io/apimachinery/pkg/api/errors"
vendor/k8s.io/client-go/util/connrotation/connrotation_test.go:103:27: should use time.Since instead of time.Now().Sub (S1012)
vendor/k8s.io/client-go/util/jsonpath/parser.go:355:7: should omit comparison to bool constant, can be simplified to !begin (S1002)
vendor/k8s.io/client-go/util/workqueue/metrics.go:185:7: receiver name should not be an underscore, omit the name if it is unused (ST1006)
vendor/k8s.io/client-go/util/workqueue/metrics.go:189:7: receiver name should not be an underscore, omit the name if it is unused (ST1006)
vendor/k8s.io/client-go/util/workqueue/metrics.go:193:7: receiver name should not be an underscore, omit the name if it is unused (ST1006)
vendor/k8s.io/client-go/util/workqueue/metrics.go:197:7: receiver name should not be an underscore, omit the name if it is unused (ST1006)
vendor/k8s.io/client-go/util/workqueue/metrics.go:201:7: receiver name should not be an underscore, omit the name if it is unused (ST1006)
vendor/k8s.io/client-go/util/workqueue/metrics.go:205:7: receiver name should not be an underscore, omit the name if it is unused (ST1006)
vendor/k8s.io/client-go/util/workqueue/metrics.go:209:7: receiver name should not be an underscore, omit the name if it is unused (ST1006)
```

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go [hunk 1]
```diff
@@ -23,6 +23,7 @@ import (
 	"errors"
 	"fmt"
 	"io"
+	"k8s.io/klog"
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:26]: Please remove all the "k8s.io/klog" lines. It's mentioned in the staticcheck result, and we already have "k8s.io/klog/v2"

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go [hunk 2]
```diff
@@ -479,7 +480,7 @@ func (a *Authenticator) refreshCredsLocked(r *clientauthentication.Response) err
 //
 // It must be called while holding the Authenticator's mutex.
 func (a *Authenticator) wrapCmdRunErrorLocked(err error) error {
-	switch err.(type) {
+	switch err := err.(type) {
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:483]: Hmm this is weird. Does L501 still work with this assignment?

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go [hunk 3]
```diff
@@ -300,7 +301,7 @@ func (c *commandTokenSource) Token() (*oauth2.Token, error) {
 	cmd.Stderr = &stderr
 	output, err := cmd.Output()
 	if err != nil {
-		return nil, fmt.Errorf("error executing access token command %q: err=%v output=%s stderr=%s", fullCmd, err, output, string(stderr.Bytes()))
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go:303]: I don't see this mentioned in the staticcheck report you pasted. Did I miss something?

File: staging/src/k8s.io/client-go/rest/config_test.go [hunk 4]
```diff
@@ -277,10 +277,10 @@ func (n *fakeNegotiatedSerializer) DecoderToVersion(serializer runtime.Decoder,
 }
 
 var fakeDialFunc = func(ctx context.Context, network, addr string) (net.Conn, error) {
-	return nil, fakeDialerError
+	return nil, ErrorFakeDialer
 }
 
-var fakeDialerError = errors.New("fakedialer")
+var ErrorFakeDialer = errors.New("fakedialer")
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/rest/config_test.go:283]: > error var fakeDialerError should have name of the form errFoo

maybe `errFakeDialer`. A variable starting with a capital letter is public, which is not necessary here

File: staging/src/k8s.io/client-go/rest/watch/decoder_test.go [hunk 5]
```diff
@@ -54,7 +54,7 @@ func TestDecoder(t *testing.T) {
 		go func() {
 			data, err := runtime.Encode(scheme.Codecs.LegacyCodec(v1.SchemeGroupVersion), expect)
 			if err != nil {
-				t.Fatalf("Unexpected error %v", err)
+				t.Errorf("Unexpected error %v", err)
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/rest/watch/decoder_test.go:57]: > vendor/k8s.io/client-go/rest/watch/decoder_test.go:54:3: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)

this is not the right fix. A proper fix may require surfacing the error back to the main goroutine through a channel.

File: staging/src/k8s.io/client-go/rest/watch/decoder_test.go [hunk 6]
```diff
@@ -70,7 +70,7 @@ func TestDecoder(t *testing.T) {
 		go func() {
 			action, got, err := decoder.Decode()
 			if err != nil {
-				t.Fatalf("Unexpected error %v", err)
+				t.Errorf("Unexpected error %v", err)
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/rest/watch/decoder_test.go:73]: similar here

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go (hunk 2 above)
Comment (roycaihw) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:482]: I see how it works now. TIL

File: staging/src/k8s.io/client-go/rest/connection_test.go [hunk 7]
```diff
@@ -43,6 +43,7 @@ type tcpLB struct {
 }
 
 func (lb *tcpLB) handleConnection(in net.Conn, stopCh chan struct{}) {
+	errs := make(chan error)
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/rest/connection_test.go:46]: this is not surfacing the error to the parent goroutine, and will block the send below. I'm afraid a proper fix is more complex than the rest of the PR is: 
- the parent goroutine needs to receive from the channel, to properly call `t.Fatalf`
- we need to be careful about when the goroutines return. If a parent goroutine returns before a child goroutine finishes / sends to the channel, the error will be ignored.
I wonder if we should leave this staticcheck error as it is and fix it in a separate PR.

File: staging/src/k8s.io/client-go/rest/request.go [hunk 8]
```diff
@@ -512,11 +512,15 @@ func (r Request) finalURLTemplate() url.URL {
 	r.params = newParams
 	url := r.URL()
 
+	if url == nil {
+		return *url
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/rest/request.go:516]: perhaps
```
if url == nil {
    klog.V(4).Infof("unexpected nil URL in Request: %v", r)
    return url.URL{}
}
```

File: staging/src/k8s.io/client-go/rest/request.go [hunk 9]
```diff
@@ -545,7 +549,7 @@ func (r Request) finalURLTemplate() url.URL {
 		// this should not happen that the only two possibilities are /api... and /apis..., just want to put an
 		// outlet here in case more API groups are added in future if ever possible:
 		// https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-groups
-		// if a wrong API groups name is encountered, return the {prefix} for url.Path
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/rest/request.go:548]: `url.Path` looks correct here

File: staging/src/k8s.io/client-go/rest/watch/decoder_test.go [hunk 10]
```diff
@@ -93,6 +95,7 @@ func TestDecoder(t *testing.T) {
 		}()
 		<-done
 
+		close(errs)
```
Comment (roycaihw) [inline staging/src/k8s.io/client-go/rest/watch/decoder_test.go:98]: similar here. We should receive from `errs` and properly call `t.Fatalf`

**Review by roycaihw — COMMENTED** [review]
I left some comments. LGTM otherwise

File: staging/src/k8s.io/client-go/rest/connection_test.go (hunk 7 above)
Comment (JaredTan95) [reply staging/src/k8s.io/client-go/rest/connection_test.go:46]: Okay , I will leave this in another PR. But, if that, I don't konw `vendor/k8s.io/client-go/rest` and `vendor/k8s.io/client-go/rest/watch` should be removed from `hack/.staticcheck_failures`?

File: staging/src/k8s.io/client-go/rest/request.go (hunk 9 above)
Comment (JaredTan95) [reply staging/src/k8s.io/client-go/rest/request.go:548]: It's a mis-change.

File: staging/src/k8s.io/client-go/rest/connection_test.go (hunk 7 above)
Comment (roycaihw) [reply staging/src/k8s.io/client-go/rest/connection_test.go]: sadly yes.. but if you could paste the new static check result and show that now it only contains this one failure, I think that's a great achievement.

**Comment by roycaihw** [comment]
/lgtm

Please squash the commits

**Comment by JaredTan95** [comment]
> squash the commits

done

File: staging/src/k8s.io/client-go/rest/watch/decoder_test.go (hunk 10 above)
Comment (roycaihw) [inline staging/src/k8s.io/client-go/rest/watch/decoder_test.go:98]: this is similar to the other place we discussed before. We cannot close the channel without calling `t.Fatalf`. Better be fixed in a separate PR.

**Review by roycaihw — COMMENTED** [review]
one comment. LGTM otherwise

File: staging/src/k8s.io/client-go/rest/watch/decoder_test.go (hunk 10 above)
Comment (JaredTan95) [reply staging/src/k8s.io/client-go/rest/watch/decoder_test.go]: got it.

**Comment by JaredTan95** [comment]
CC @roycaihw

**Comment by roycaihw** [comment]
/lgtm

Thanks!

**Comment by neolit123** [comment]
/remove-area kubeadm
/remove-sig cluster-lifecycle

please update the PR to only include the areas mentioned in the description.

**Comment by BenTheElder** [comment]
this PR needs a rebase, the actual changes are obscured by the PR showing up as having more than 700 commits.

**Comment by roycaihw** [comment]
let me know when you clean up the commits and I will re-tag the PR
